File: java/src/main/java/com/workshop/architecture/fitness/managing_memberships/shared/MembershipEntity.java
package com.workshop.architecture.fitness.managing_memberships.shared;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.UUID;

@Entity
@Table(name = "memberships")
public class MembershipEntity {

    @Id
    private UUID id;

    @Column(nullable = false)
    private String customerId;

    @Column(nullable = false)
    private String planId;

    @Column(nullable = false)
    private int planPrice;

    @Column(nullable = false)
    private int planDuration;

    @Column(nullable = false)
    private String status;

    @Column
    private String reason;

    @Column(nullable = false)
    private LocalDate startDate;

    @Column(nullable = false)
    private LocalDate endDate;

    @Column
    private LocalDate pauseStartDate;

    @Column
    private LocalDate pauseEndDate;

    @Column
    private String pauseReason;

    @Column
    private Instant cancelledAt;

    @Column
    private String cancellationReason;

    protected MembershipEntity() {
    }

    public MembershipEntity(
            UUID id,
            String customerId,
            String planId,
            int planPrice,
            int planDuration,
            String status,
            String reason,
            LocalDate startDate,
            LocalDate endDate
    ) {
        this.id = id;
        this.customerId = customerId;
        this.planId = planId;
        this.planPrice = planPrice;
        this.planDuration = planDuration;
        this.status = status;
        this.reason = reason;
        this.startDate = startDate;
        this.endDate = endDate;
    }

    public UUID getId() {
        return id;
    }

    public String getCustomerId() {
        return customerId;
    }

    public String getPlanId() {
        return planId;
    }

    public int getPlanPrice() {
        return planPrice;
    }

    public int getPlanDuration() {
        return planDuration;
    }

    public String getStatus() {
        return status;
    }

    public String getReason() {
        return reason;
    }

    public LocalDate getStartDate() {
        return startDate;
    }

    public LocalDate getEndDate() {
        return endDate;
    }

    public LocalDate getPauseStartDate() {
        return pauseStartDate;
    }

    public LocalDate getPauseEndDate() {
        return pauseEndDate;
    }

    public String getPauseReason() {
        return pauseReason;
    }

    public Instant getCancelledAt() {
        return cancelledAt;
    }

    public String getCancellationReason() {
        return cancellationReason;
    }

    public boolean isActive() {
        return "ACTIVE".equals(status);
    }

    public boolean isPaused() {
        return "PAUSED".equals(status);
    }

    public boolean isSuspended() {
        return "SUSPENDED".equals(status);
    }

    public boolean isSuspendedForNonPayment() {
        return isSuspended() && "NON_PAYMENT".equals(reason);
    }

    // Is this domain or data access?
    public boolean isCancelled() {
        return "CANCELLED".equals(status);
    }

    public void suspend() {
        this.status = "SUSPENDED";
    }

    public void suspendForNonPayment() {
        this.status = "SUSPENDED";
        this.reason = "NON_PAYMENT";
    }

    public void reactivateAfterPayment() {
        this.status = "ACTIVE";
        this.reason = null;
    }

    public void pause(LocalDate pauseStartDate, LocalDate pauseEndDate, String reason) {
        long pauseDays = ChronoUnit.DAYS.between(pauseStartDate, pauseEndDate) + 1;
        this.status = "PAUSED";
        this.pauseStartDate = pauseStartDate;
        this.pauseEndDate = pauseEndDate;
        this.pauseReason = reason;
        this.endDate = this.endDate.plusDays(pauseDays);
    }

    public void resumeAfterPause() {
        this.status = "ACTIVE";
        this.pauseStartDate = null;
        this.pauseEndDate = null;
        this.pauseReason = null;
    }

    public void cancel(Instant cancelledAt, String reason) {
        this.status = "CANCELLED";
        this.cancelledAt = cancelledAt;
        this.cancellationReason = reason;
    }

    public void extendBy(int additionalMonths, int additionalDays) {
        this.endDate = this.endDate.plusMonths(additionalMonths).plusDays(additionalDays);
    }
}

File: python/src/workshop_api/fitness/shared/membership_entity.py
from __future__ import annotations

import uuid
from datetime import datetime, timedelta

from sqlalchemy import Column, Date, DateTime, Integer, String

from workshop_api.fitness.database import Base


class MembershipOrmModel(Base):
    __tablename__ = "memberships"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    customer_id = Column(String(36), nullable=False)
    plan_id = Column(String(36), nullable=False)
    plan_price = Column(Integer, nullable=False)
    plan_duration = Column(Integer, nullable=False)
    status = Column(String(50), nullable=False)
    reason = Column(String(100), nullable=True)
    start_date = Column(Date, nullable=False)
    end_date = Column(Date, nullable=False)
    pause_start_date = Column(Date, nullable=True)
    pause_end_date = Column(Date, nullable=True)
    pause_reason = Column(String(100), nullable=True)
    cancelled_at = Column(DateTime(timezone=True), nullable=True)
    cancellation_reason = Column(String(100), nullable=True)

    def is_active(self) -> bool:
        return self.status == "ACTIVE"

    def is_paused(self) -> bool:
        return self.status == "PAUSED"

    def is_suspended(self) -> bool:
        return self.status == "SUSPENDED"

    def is_suspended_for_non_payment(self) -> bool:
        return self.is_suspended() and self.reason == "NON_PAYMENT"

    def is_cancelled(self) -> bool:
        return self.status == "CANCELLED"

    def suspend_for_non_payment(self) -> None:
        self.status = "SUSPENDED"
        self.reason = "NON_PAYMENT"

    def reactivate_after_payment(self) -> None:
        self.status = "ACTIVE"
        self.reason = None

    def pause(self, pause_start_date, pause_end_date, reason: str | None) -> None:
        pause_days = (pause_end_date - pause_start_date).days + 1
        self.status = "PAUSED"
        self.pause_start_date = pause_start_date
        self.pause_end_date = pause_end_date
        self.pause_reason = reason
        self.end_date = self.end_date + timedelta(days=pause_days)

    def resume_after_pause(self) -> None:
        self.status = "ACTIVE"
        self.pause_start_date = None
        self.pause_end_date = None
        self.pause_reason = None

    def cancel(self, cancelled_at, reason: str | None) -> None:
        self.status = "CANCELLED"
        self.cancelled_at = cancelled_at
        self.cancellation_reason = reason

    def extend_by(self, additional_months: int, additional_days: int) -> None:
        month_index = self.end_date.month - 1 + additional_months
        year = self.end_date.year + month_index // 12
        month = month_index % 12 + 1
        import calendar

        day = min(self.end_date.day, calendar.monthrange(year, month)[1])
        self.end_date = self.end_date.replace(year=year, month=month, day=day)
        self.end_date = self.end_date + timedelta(days=additional_days)


class MembershipBillingReferenceOrmModel(Base):
    __tablename__ = "membership_billing_references"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    membership_id = Column(String(36), nullable=False)
    external_invoice_id = Column(String(100), nullable=False, unique=True)
    external_invoice_reference = Column(String(100), nullable=False, unique=True)
    due_date = Column(Date, nullable=False)
    status = Column(String(50), nullable=False)
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

    def is_paid(self) -> bool:
        return self.status == "PAID"

    def mark_paid(self, paid_at) -> None:
        self.status = "PAID"
        self.updated_at = paid_at
